java/google/registry/rde/imports/RdeHostLinkAction.java

```diff
@@ -0,0 +1,162 @@
+    this.importBucketName = importBucketName;
+    this.importFileName = importFileName;
+    this.mapShards = mapShards;
+  }
+
+  @Override
+  public void run() {
+    response.sendJavaScriptRedirect(createJobPath(mrRunner
+        .setJobName("Link hosts from escrow file")
+        .setModuleName("backend")
+        .runMapOnly(
+            new RdeHostPostImportMapper(),
+            ImmutableList.of(new RdeHostInput(mapShards, importBucketName, importFileName)))));
+  }
+
+  /**
+   * Return the {@link DomainResource} this host is subordinate to, or null for out of zone hosts.
+   *
+   * @throws {@link IllegalStateException} for hosts without superordinate domains
+   */
+  @Nullable
+  private static DomainResource lookupSuperordinateDomain(InternetDomainName hostName, DateTime now) {
+    Optional<InternetDomainName> tld = findTldForName(hostName);
+    // out of zone hosts cannot be linked
+    if(!tld.isPresent()) {
+      return null;
+    }
+    // This is a subordinate host
+    String domainName = Joiner.on('.').join(Iterables.skip(
+        hostName.parts(), hostName.parts().size() - (tld.get().parts().size() + 1)));
+    DomainResource superordinateDomain = loadByForeignKey(DomainResource.class, domainName, now);
+    // Hosts can't be linked if domains import hasn't been run
+    checkState(superordinateDomain != null,
+      "Superordinate domain does not exist: %s",
+      domainName);
+    return superordinateDomain;
+  }
+
+  /** Mapper to link hosts from an escrow file to their superordinate domains. */
+  public static class RdeHostPostImportMapper extends Mapper<JaxbFragment<XjcRdeHostElement>, Void, Void> {
+
+    private static final long serialVersionUID = -2898753709127134419L;
+
+    @Override
+    public void map(JaxbFragment<XjcRdeHostElement> fragment) {
+      // Record number of attempted map operations
+      getContext().incrementCounter("post-import hosts read");
+      final XjcRdeHost xjcHost = fragment.getInstance().getValue();
+      logger.infofmt("Attempting to link superordinate domain for host %s", xjcHost.getName());
+      try {
+        InternetDomainName hostName = InternetDomainName.from(xjcHost.getName());
+        DomainResource superordinateDomain = lookupSuperordinateDomain(hostName, DateTime.now());
+        // if suporordinateDomain is null, this is an out of zone host and can't be linked
+        if (superordinateDomain == null) {
+          getContext().incrementCounter("post-import hosts out of zone");
+          logger.infofmt("Host %s is out of zone", xjcHost.getName());
+          return;
+        }
+        // at this point, the host is definitely in zone and should be linked
+        getContext().incrementCounter("post-import hosts in zone");
+        final Key<DomainResource> superordinateDomainKey = Key.create(superordinateDomain);
+        ofy().transact(new VoidWork() {
+          @Override
+          public void vrun() {
+            HostResource host =
```

Comment:
Is it ever possible that the host will fail to load at this point?  If it does, you're going to NPE in the line below when you call host.asBuilder().  It might be worth handling just to get a more useful fatal error message in the logs.